dotCMS/src/main/java/com/dotcms/security/apps/AppsCacheImpl.java

```diff
@@ -0,0 +1,213 @@
+        }
+        return descriptorsByKey;
+    }
+
+    /**
+     * This simply allow putting in cache
+     * A Map like structure, Where the key is the app-key and the entry is the AppDescriptor.
+     * @param descriptorsByKey
+     */
+    private void putDescriptorsByKey(final Map<String, AppDescriptorMeta> descriptorsByKey){
+        cache.put(DESCRIPTORS_MAPPED_BY_KEY, descriptorsByKey, DESCRIPTORS_CACHE_GROUP);
+    }
+
+    /**
+     * Remove all descriptors from cache;
+     */
+    public void invalidateDescriptorsCache(){
+        cache.flushGroup(DESCRIPTORS_CACHE_GROUP);
+    }
+
+    /**
+     * In a similar fashion this takes a supplier delegate responsible for getting the secret keys to populate cache.
+     * If the supplier is null the function will get back the contents from cache and no population attempt will be performed.
+     * @param supplier string keys set delegate
+     * @return
+     * @throws DotCacheException
+     */
+    public synchronized Set<String> getKeysFromCache(final Supplier<Set<String>> supplier) throws DotCacheException {
+        Set<String> keys = (Set<String>) cache.get(SECRETS_CACHE_KEY, SECRETS_CACHE_KEYS_GROUP);
+        if (!UtilMethods.isSet(keys) && null != supplier) {
+            keys = supplier.get();
+            putKeys(keys);
+        }
+        return keys;
+    }
+
+    /**
+     * In this version of the method the supplier is null
+     * so the function will get back the contents from cache and no population attempt will be performed.
+     * @return
+     * @throws DotCacheException
+     */
+    public Set<String> getKeysFromCache() throws DotCacheException {
+       return getKeysFromCache(null);
+    }
+
+    /**
+     * Given a supplier delegate able to retrieve an encrypted secret this method will use it to populate cache.
+     * @param key Key
+     * @param supplier encrypted secret supplier
+     * @return encrypted secret from cache.
+     */
+    public synchronized char[] getFromCache(final String key, final Supplier<char[]> supplier) {
+        char[] retVal = (char[]) cache.getNoThrow(key, SECRETS_CACHE_GROUP);
+        if (retVal == null && null != supplier) {
+            retVal = supplier.get();
+            putSecret(key, retVal);
+        }
+        return retVal;
+    }
+
+    /**
+     * In this version of the method the supplier is null
+     * so the function will get back the secrets from cache and no population attempt will be performed.
+     * @param key secret key
+     * @return encrypted secret from cache.
+     */
+    public synchronized char[] getFromCache(final String key){
+       return getFromCache(key, null);
+    }
+
+    /**
+     * Simply Puts a secret in cache.
+     * @param key
+     * @param chars
+     */
+    public void putSecret(final String key, final char[] chars){
+        cache.put(key, chars, SECRETS_CACHE_GROUP);
+    }
+
+    /**
+     * All Secrets flush.
+     */
+    public synchronized void flushSecret() {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use block level rather than method level synchronization](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=5358201)